### PR TITLE
Porting EDK II to Allwinner A13 tablet

### PR DESCRIPTION
**Pull request by arturkow2000**

### CODE REVIEW

File: blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+image. Implementing USB driver makes it possible to load EFI directly from Pi
+without going back into BootROM, thus allowing to move initialization code
+there. Implementing it also enables UEFI to be debugged using USB instead of
+UART.
+
+## Supported hardware
+
+Currently, only my XW711 tablet is supported, but this hardware is so similar
+to Q8 that it should work on any Q8 tablet, these look like the same board, with
+only some peripherals like touch panel and LCD differing.
```
Comment (pietrushnic) [inline blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md:60]: @arturkow2000 more precise links to specs of those devices + description what we dealing with here (SoC name, peripherals, memory, connectivity, screen size etc.) would be appreciated.

File: blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+- CPU frequency scaling
+- LCD
+- HDMI
+- DMA
+- UEFI variables - currently, variables are not preserved across reboots
+- ACPI
+- Support for other CPUs and boards.
+
+Currently, I am working on USB peripheral mode support. It will allow making
+boot process compliant with Pi specification. Currently, a small executable
```
Comment (pietrushnic) [inline blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md:48]:
```suggestion
boot process compliant with PI specification. Currently, a small executable
```

File: blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+Since there is no display support and USB is still on its way, UART connection
+is required, on this board (and on Q8), UART pads are located on the back of the
+board, there is also a second UART multiplexed with µSD, which can be used
+without disassembling device, support for this is coming soon.
+
+## Building and booting UEFI
+
+Source code is available
+[here](https://github.com/arturkow2000/SunxiPlatformPkg), you can build it using
+`arm-none-eabi-gcc`.
```
Comment (pietrushnic) [inline blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md:71]: @arturkow2000 where to get it and what version was tested?

Comment (pietrushnic) [reply blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md:71]: maybe using 3mdeb docker for edk2 would be more scalable and less dependent on readers OS

File: blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+* DTB must be manually loaded from file on each boot, using `setfdt` EFI shell
+  command
+
+* I had to boot Linux with `cpufreq.off=1`, or it would hang while trying to
+  raise CPU frequency (since there is no frequency scaling CPU is left at 384
+  MHz, this looks like Linux bug)
+
+* I had to disable `axp20x_adc` driver because it was causing board to power off
+  instantly, this isn't related to EFI itself, yet still, it's causing problems.
+  It can be caused by a bug in DTB or driver itself, or XW711 and Q8 aren't o
```
Comment (pietrushnic) [inline blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md:110]: o?

File: blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  similar, and I'm just using wrong DTB.
+```shell
+echo 'blacklist axp20x_adc' >> /etc/modprobe.d/blacklist.conf
+```
```
Comment (pietrushnic) [inline blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md:114]: @arturkow2000 standard blog summary is missing

**Review by pietrushnic — CHANGES_REQUESTED** [review]
@arturkow2000 I appreciate your writing, it can be largely extended and even should be. Please imrove abstract to include more information from intro, please note abstract is announced on social media.

Anyway great job, we need more beefy content like this. Feel free to add things you tried to get to current state, what is your next priority in this project, why you choose this approach vs the others.

File: blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md [hunk 6]
```diff
@@ -0,0 +1,168 @@
+---
+title: Porting EDK II to an old Allwinner A13 tablet
+abstract: 'An effort to port EDK II to Allwinner SoCs'
+cover: /covers/image-file.png
```
[the remark this answers is not in the document]
Comment (arturkow2000) [reply blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md]: @arturkow2000 I brought back old display display and I put here a photo of tablet running UEFI. I have also updated post with two images of tablet running UEFI and GRUB.

File: blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md [hunk 7]
```diff
@@ -0,0 +1,168 @@
+---
+title: Porting EDK II to an old Allwinner A13 tablet
+abstract: 'An effort to port EDK II to Allwinner SoCs'
```
[the remark this answers is not in the document]
Comment (arturkow2000) [reply blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md]: @pietrushnic Does it look OK now?

File: blog/content/authors/artur-kowalski/_index.md [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+identifier: artur.kowalski
+photo: "/authors/artur.kowalski.jpg"
+facebook: ""
+twitter: ""
+linkedin: ""
+github: "https://github.com/arturkow2000"
+---
+Junior embedded developer at 3mdeb. Interested in low-level development ranging
+from microcontroller programming to hypervisor and kernel development. In free
+time working on various personal open-source projects.
```
Comment (arturkow2000) [inline blog/content/authors/artur-kowalski/_index.md:12]: @pietrushnic I added bio and photo to my profile.

File: blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+
+- DTB must be manually loaded from file on each boot, using `setfdt` EFI shell
+  command
+
+- I had to boot Linux with `cpufreq.off=1`, or it would hang while trying to
+  raise CPU frequency (since there is no frequency scaling, CPU is left at 384
+  MHz), this looks like Linux's bug.
+
+- I had to disable `axp20x_adc` driver because it was causing board to power off
+  instantly, this isn't related to EFI itself, yet still, it's causing problems.
```
Comment (pietrushnic) [inline blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md:167]: Let's be consistent about naming UEFI is the name of specification. EFI is pre-UEFI non-public specification, so I doubt you using it. To be precise we can use UEFI firmware or just firmware, boot firmware even EDK II.

File: blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+```shell
+echo 'blacklist axp20x_adc' >> /etc/modprobe.d/blacklist.conf
+```
+
+{{< youtube PjRC6vXxlpY >}}
+
+## Summary
+
+Most of the basic stuff is already there. There is still a lot of work to do to
+complete work, but it is even now almost functional. To bring this into a usable
```
Comment (pietrushnic) [inline blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md:179]: it, this and other not precise wording

**Review by pietrushnic — CHANGES_REQUESTED** [review]
@arturkow2000 I really like what you did so far. I'm sorry to nitpicking more, but this is normal practice while writing first time. Please do not be discouraged I just want to improve your style and make good marketing material for 3mdeb - from perspective of time I'm pretty sure you will be proud of your first blog post.

File: blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+## Supported hardware
+
+Currently, only my XW711 tablet is supported, but this hardware is so similar
+to Q8 that it should work on any Q8 tablet, these look like the same board, with
+only some peripherals like touch panel and LCD differing.
+
+Information about Q8 is available at [sunxi wiki](https://linux-sunxi.org/Q8),
+there are no significant information available for XW711, except in its FEX file
+available in [sunxi-boards](https://github.com/linux-sunxi/sunxi-boards/blob/master/sys_config/a13/szenio_1207c4.fex)
+repo.
```
[the remark this answers is not in the document]
Comment (arturkow2000) [reply blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md]: @pietrushnic Information about Q8 is already available. It's XW711 what is missing.

File: blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+- UEFI variables - currently, variables are not preserved across reboots
+- ACPI
+- SD/MMC write support
+- Support for other CPUs and boards.
+
+Currently, I am working on USB peripheral mode support. It will allow making
+boot process compliant with PI specification. Currently, a small executable
+(called SPL after U-Boot's SPL) is loaded into the device's memory, it
+initializes DRAM and returns into BootROM, which proceeds into loading UEFI
+image. Implementing USB driver makes it possible to load EFI directly from PI
```
[the remark this answers is not in the document]
Comment (arturkow2000) [reply blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md]: @pietrushnic BootROM loads full FD file over USB. It's not that BootROM has some features to UEFI images, it has support for writing arbitrary data to memory. I just write raw FD file to some memory address and jump to it.

File: blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+- CPU frequency scaling
+- HDMI
+- DMA
+- UEFI variables - currently, variables are not preserved across reboots
+- ACPI
+- SD/MMC write support
+- Support for other CPUs and boards.
+
+Currently, I am working on USB peripheral mode support. It will allow making
+boot process compliant with PI specification. Currently, a small executable
```
[the remark this answers is not in the document]
Comment (arturkow2000) [reply blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md]: @pietrushnic BootROM can't load full UEFI image into RAM, so to load it I execute SPL which initializes DRAM and returns to BootROM, then I can upload the rest of UEFI.
DRAM initialization should be done from PEI and I can't go back from PEI to BootROM because that wouldn't be compliant with PI spec.

I'm not even sure whether it's possible to make boot process compliant at all - PI spec says:
> The Security (SEC) phase is the first phase in the PI Architecture architecture and is responsible for
> the following:
> • Handling all platform restart events
> • Creating a temporary memory store
> • Serving as the root of trust in the system
> • Passing handoff information to the PEI Foundation
But most ARM SoCs come with BootROM, which is the first phase.

Also BootROM has 31 KiB limit on A13 and even less on A10 and A20 (SRAM is much bigger than that). PEI may not fit. I can bypass this limit when loading over USB but not when loading from MMC, so probably I will still need some separate program just to load PEI.

I removed PI compliance from blogpost. Still I want to move DRAM initialization into PEI because doing so will allow me to fetch DRAM config from dynamic PCDs, which will be useful for overclocking.

File: blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (arturkow2000) [reply blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md]: @pietrushnic I will do that when post cames out.

File: blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+(called SPL after U-Boot's SPL) is loaded into the device's memory, it
+initializes DRAM and returns into BootROM, which proceeds into loading UEFI
+image. Implementing USB driver makes it possible to load EFI directly from PI
+without going back into BootROM, thus allowing to move initialization code
+there. Implementing it also enables UEFI to be debugged using USB instead of
+UART.
```
[the remark this answers is not in the document]
Comment (arturkow2000) [reply blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md]: @pietrushnic I avoided these as much as I could, I hope it looks good now.

File: blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (pietrushnic) [reply blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md]: All the above concerns are IMO valid and should be in your blog post. It is very good question to ask "what the heck it even mean to be PI spec compliant?" - it is very hard to follow vague spec. More to that I think SEC phase was described in such way to make it as flexible as possible, because PI spec does not want to limit its adoption around platform initialization. Key point here is that PEI is bloated and system you describe has hard time with resources. Whole situation creates great opportunity for true embedded firmware development.

Now I understand why you load from USB, but I think our above discussion should in some form be introduced into the blog post.

File: blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (pietrushnic) [reply blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md]: @arturkow2000 please add that explanation to blog post.

**Comment by pietrushnic** [comment]
@arturkow2000 we have 2 last unresolved discussions before publication.

File: blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (arturkow2000) [reply blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md]: done

File: blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (arturkow2000) [reply blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md]: @pietrushnic I extended post with some information, notably, additional benefits and the problem with PEI size, still I'm thinking about PI compliance part.

**Comment by pietrushnic** [comment]
@arturkow2000 force-push does not help much with checking what changed :D @artur-rs maybe some policy about force pushes as it is in Dasharo: https://docs.dasharo.com/dev-proc/source-code-structure/#force-pushes-rules

File: blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md [hunk 15]
```diff
@@ -77,9 +77,19 @@ feature in UEFI.
...
+BootROM and the need for SPL. Moving DRAM initialization code into PEI makes it
+possible to fetch DRAM configuration from dynamic PCDs, which will be useful for
+overclocking. Yet another significant advantage is that I can use the PEI MMC
+driver to load the rest of UEFI, so when I get rid of SPL, I will add support
+for UEFI booting from SD and eMMC. There is only one problem with this approach,
+BootROM has a 31 KiB limit on A13 and even less on A10 and A20 (despite SRAM
+being much bigger than that). Even now, PEI size exceeds that limit. There is a
+chance that I could compact it enough to make it fit, but I can't be sure, and
+if it turns impossible, then I will have to keep a small MMC driver in SEC just
+to load PEI.
```
Comment (pietrushnic) [inline blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md:92]: @arturkow2000 I like this explanation a lot. Thank you. Keep the good work.

File: blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (arturkow2000) [reply blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md]: @pietrushnic I added few words about PI compliance

**Review by pietrushnic — APPROVED** [review]
